builtin/v10/check.go

```diff
@@ -0,0 +1,367 @@
+package v10
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/filecoin-project/go-state-types/builtin/v10/datacap"
+	"github.com/filecoin-project/go-state-types/builtin/v10/evm"
+	"github.com/filecoin-project/go-state-types/manifest"
+	"github.com/ipfs/go-cid"
+
+	"github.com/filecoin-project/go-state-types/builtin/v10/account"
+	"github.com/filecoin-project/go-state-types/builtin/v10/cron"
+	init_ "github.com/filecoin-project/go-state-types/builtin/v10/init"
+	"github.com/filecoin-project/go-state-types/builtin/v10/market"
+	"github.com/filecoin-project/go-state-types/builtin/v10/miner"
+	"github.com/filecoin-project/go-state-types/builtin/v10/multisig"
+	"github.com/filecoin-project/go-state-types/builtin/v10/paych"
+	"github.com/filecoin-project/go-state-types/builtin/v10/power"
+	"github.com/filecoin-project/go-state-types/builtin/v10/reward"
+	"github.com/filecoin-project/go-state-types/builtin/v10/verifreg"
+
+	"github.com/filecoin-project/go-state-types/builtin"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/big"
+	"golang.org/x/xerrors"
+)
+
+// Within this code, Go errors are not expected, but are often converted to messages so that execution
+// can continue to find more errors rather than fail with no insight.
+// Only errors thar are particularly troublesome to recover from should propagate as Go errors.
+func CheckStateInvariants(tree *builtin.ActorTree, priorEpoch abi.ChainEpoch, actorCodes map[string]cid.Cid) (*builtin.MessageAccumulator, error) {
+	acc := &builtin.MessageAccumulator{}
+	totalFIl := big.Zero()
+	var initSummary *init_.StateSummary
+	var cronSummary *cron.StateSummary
+	var verifregSummary *verifreg.StateSummary
+	var datacapSummary *datacap.StateSummary
+	var marketSummary *market.StateSummary
+	var rewardSummary *reward.StateSummary
+	var accountSummaries []*account.StateSummary
+	var powerSummary *power.StateSummary
+	var paychSummaries []*paych.StateSummary
+	var multisigSummaries []*multisig.StateSummary
+	var delegatedAddrs []address.Address
+	minerSummaries := make(map[address.Address]*miner.StateSummary)
+
+	emptyObjectCid, err := builtin.MakeEmptyState()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := tree.ForEachV5(func(key address.Address, actor *builtin.ActorV5) error {
+		acc := acc.WithPrefix("%v ", key) // Intentional shadow
+		if key.Protocol() != address.ID {
+			acc.Addf("unexpected address protocol in state tree root: %v", key)
+		}
+		totalFIl = big.Add(totalFIl, actor.Balance)
+
+		if actor.Address != nil && actor.Address.Protocol() == address.Delegated {
```

Comment:
I'd slightly tweak this: If the address is non-nil, it MUST be a delegated address, so assert that instead of checking-and-skipping.